packages/jsii-kernel/lib/kernel.ts

```diff
@@ -756,83 +746,90 @@ export class Kernel {
+      for (const base of bases) {
+        if (!base) { continue; }
 
-      const found = this._tryTypeInfoForMethod(base, methodName);
-      if (found) {
-        return found;
+        const found = this._tryTypeInfoForMethod(methodName, base);
+        if (found) {
+          return found;
+        }
       }
     }
 
     return undefined;
   }
 
-  private _tryTypeInfoForProperty(fqn: string, property: string): spec.Property | undefined {
-    if (!fqn) {
-      throw new Error('missing "fqn"');
-    }
-    const typeInfo = this._typeInfoForFqn(fqn);
-
-    let properties;
-    let bases;
-
-    if (spec.isClassType(typeInfo)) {
-      const classTypeInfo = typeInfo as spec.ClassType;
-      properties = classTypeInfo.properties;
-      bases = classTypeInfo.base ? [classTypeInfo.base] : [];
-    } else if (spec.isInterfaceType(typeInfo)) {
-      const interfaceTypeInfo = typeInfo as spec.InterfaceType;
-      properties = interfaceTypeInfo.properties;
-      bases = interfaceTypeInfo.interfaces || [];
-    } else {
-      throw new Error(`Type of kind ${typeInfo.kind} does not have properties`);
-    }
+  private _tryTypeInfoForProperty(property: string, classFqn: string, interfaces: string[] = []): spec.Property | undefined {
+    for (const fqn of [classFqn, ...interfaces]) {
+      if (fqn === wire.EMPTY_OBJECT_FQN) { continue; }
+      const typeInfo = this._typeInfoForFqn(fqn);
+
+      let properties;
+      let bases;
+
+      if (spec.isClassType(typeInfo)) {
+        const classTypeInfo = typeInfo as spec.ClassType;
+        properties = classTypeInfo.properties;
+        bases = classTypeInfo.base ? [classTypeInfo.base] : [];
+      } else if (spec.isInterfaceType(typeInfo)) {
+        const interfaceTypeInfo = typeInfo as spec.InterfaceType;
+        properties = interfaceTypeInfo.properties;
+        bases = interfaceTypeInfo.interfaces || [];
+      } else {
+        throw new Error(`Type of kind ${typeInfo.kind} does not have properties`);
+      }
 
-    for (const p of properties || []) {
-      if (p.name === property) {
-        return p;
+      for (const p of properties || []) {
+        if (p.name === property) {
+          return p;
+        }
       }
-    }
 
-    // recurse to parent type (if exists)
-    for (const baseFqn of bases) {
-      const ret = this._tryTypeInfoForProperty(baseFqn, property);
-      if (ret) {
-        return ret;
+      // recurse to parent type (if exists)
+      for (const baseFqn of bases) {
+        const ret = this._tryTypeInfoForProperty(property, baseFqn);
+        if (ret) {
+          return ret;
+        }
       }
     }
 
     return undefined;
   }
 
-  private _typeInfoForProperty(fqn: string, property: string): spec.Property {
-    const typeInfo = this._tryTypeInfoForProperty(fqn, property);
+  private _typeInfoForProperty(property: string, fqn: string, interfaces?: string[]): spec.Property {
```

Comment:
In effect, shouldn't the last 2 arguments just be a list of FQNs, which can be either class or interface types?

Reply:
There can only be ONE class there, and I feel like it's easier to represent that information if it's a separate argument. I probably should rename it though.